doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,592 @@
+    OVER (ORDER BY time PARTITION BY sensor_id)
+FROM measurements;
+```
+
+Certain window functions operate on a _window frame_, which is a subset of a partition. The default frame includes the rows from the first row of the partition up to the current row (or more accurately, to the last row of the peer group of the current row, where a peer group is a set of rows that are equal on both the `PARTITION BY` and the `ORDER BY`). For example, all aggregation functions can be used also as window functions (we will refer to this as _window aggregations_), where they aggregate values from inside the current window frame. The following query calculates a running total (prefix sum) of measurement values for each sensor (which wouldn't make sense for a temperature sensor, but makes sense for, e.g., a water flow sensor):
+
+```SQL
+SELECT sensor_id, time, SUM(measurement_value)
+    OVER (ORDER BY time PARTITION BY sensor_id)
+FROM measurements;
+```
+
+Note that this query doesn't compute just one value for each partition. Instead, it calculates a value for each input row: the sum of the same sensor's measurements that happened no later than the current input row.
+
+We can also explicitly specify a frame, i.e., how far it extends from the current row, both backwards and forwards. One option is to say `UNBOUNDED PRECEDING` or `UNBOUNDED FOLLOWING`, meaning that the frame extends to the beginning or end of the current partition. Another option is to specify an offset. For example, the following query computes a moving average (e.g., to have a smoother curve when we want to plot it or when we want less noise for an alerting use case):
+
+```SQL
+SELECT sensor_id, time, AVG(measurement_value)
+    OVER (ORDER BY time PARTITION BY sensor_id
+        ROWS BETWEEN 4 PRECEDING AND CURRENT ROW)
+FROM measurements;
+```
+
+In this query, the frame extends 4 rows backwards, and ends at the current row (thus containing at most 5 rows).
+
+The exact meaning of the offset depends on the _frame mode_:
+- In `ROWS` mode (such as above), the frame extends for the specified number of rows (or less, for rows near the beginning or end of the partition).
+- In `GROUPS` mode, the frame extends for the specified number of peer groups, where a peer group is a set of rows that are equal on both the `PARTITION BY` and the `ORDER BY`.
+- In `RANGE` mode, the frame extends to those rows whose difference from the current row on the `ORDER BY` column is not greater than the offset (only one ORDER BY column is allowed for this frame mode). For example, the following query computes a moving average with a frame size of 5 minutes (which might be more useful than a `ROWS` offset when the measurement values are at irregular times):
+```SQL
+SELECT sensor_id, time, AVG(measurement_value)
+    OVER (ORDER BY time PARTITION BY sensor_id
+        RANGE BETWEEN '5 minutes' PRECEDING AND CURRENT ROW)
+FROM measurements;
+```
+
```

Comment:
It's good that you asked this, because I realized I can't even think of a single good example... Any examples that I found have the problem that they either don't make much sense, or can be easily solved without the GROUPS frame mode. (There are a several not-so-good examples here: https://mjk.space/advances-sql-window-frames/) 

Maybe we can postpone supporting it until we see at least one actual user request.

Reply:
This is more of a product-level decision, but either way, it might be good to be explicit in the document that we are not supporting this functionality. My understanding is that we do not support it right now: https://github.com/MaterializeInc/materialize/blob/928a91fa97bea88f49b843763edd9343cf6b6f89/src/expr/src/relation/func.rs#L1006-L1007